SurveyApp/Modules/Login/LoginInteractor.swift

```diff
@@ -6,20 +6,49 @@
 //  
 //
 
+import Foundation
+
 // sourcery: AutoMockable
 protocol LoginInteractorInput: AnyObject {
+    
+    func authenticateEmail(email: String, password: String)
 }
 
 // sourcery: AutoMockable
 protocol LoginInteractorOutput: AnyObject {
+    
+    func didAuthenticateEmail()
+    func didFailToAuthenticateEmail()
 }
 
 final class LoginInteractor {
 
     weak var output: LoginInteractorOutput?
+    
+    private(set) var authenticationService: AuthenticationServiceProtocol?
+    
+    private var authenticateEmailRequest: Request? {
+        didSet { oldValue?.cancel() }
+    }
+    
+    init(authenticationService: AuthenticationServiceProtocol) {
+        self.authenticationService = authenticationService
+    }
 }
 
 // MARK: - LoginInteractorInput
 
 extension LoginInteractor: LoginInteractorInput {
+    
+    func authenticateEmail(email: String, password: String) {
+        authenticateEmailRequest = authenticationService?.authenticateEmail(email: email, password: password) { [weak output] result in
+            switch result {
+            case .success(let authToken):
+                UserSessionProvider.shared.userSession?.userCredential = authToken.session?.userCredential
```

Comment:
This could be better with dependency injection. 
```
final class LoginInteractor {
        ...
        let userSessionProvider: UserSessionProviderProtocol

        init(
               authenticationService: AuthenticationServiceProtocol,
               userSessionProvider: UserSessionProviderProtocol
          ) {
                  ...
                  self.userSessionProvider = userSessionProvider
          }

    func authenticateEmail(email: String, password: String) {
          ...
                self?.userSessionProvider.userSession?.userCredential = authToken.session?.userCredential
    }
}
```

Module.swift:
```
let userSessionProvider = UserSessionProvider.shared
... = LoginInteractor(..., userSessionProvider: userSessionProvider)
```


This will reduce number of singletons and increase testability.

Reply:
Thank you for the better approach 🙆‍♀️

Reply:
@blyscuit I updated the code please help check 🙇‍♀️